doc/source/dev/development_workflow.rst

```diff
@@ -150,6 +153,88 @@ been added to ``upstream`` that affect your work. In this case, follow the
 :ref:`rebasing-on-main` section of this document to apply those changes to
 your branch.
 
+.. _formatting-with-pre-commit:
+
+Automatic code formatting with pre-commit
+-----------------------------------------
+
+NumPy recommends that code follow PEP7/PEP8 guidelines, but does not currently
+enfoce any formatters. However, the repository does come with a an optional
+pre-commit hook that performs automatic formatting and validation of changes,
+and using it is highly recommended. Tasks and checks that it runs include:
+
+* Format changed lines only of Python files to pep8 standard using `Darker
+  <darker_>`_, which wraps `black <black_>`_ and `isort <isort_>`_ and applies
+  them to the diff (this hook is enabled by default). Additionally, check only
+  these changed lines using `flake8 <flake8_>`_.
+
+* Format changed lines only of C and C++ source files to pep7 standard using a
+  python wrapper for `ClangFormat versioning <clang_format_>`_ (this hook is
+  enabled by default).
+
+* Format and lint YAML shell files, and perform linting on Fortran, rst, Docker,
+  and markdown files (some enabled and some disabled by default)
+
+To use the hook, first make sure that pre-commit is installed:
+
+.. code-block:: bash
+
+   pip install pre-commit
+
+And then install the hook to your git repository:
+
+.. code-block:: bash
+
+   pre-commit install
+
+This will add a script to your ``.git/hooks`` folder. This script will run
+pre-commit to check your changed files whenever you run ``git commit``. Note
+that if there are any failures, the commit will not go through and you will have
+to re-run the commit (and ``git add`` first if any files were automatically
+changed, or ``git commit`` with the ``-a`` flag). It is generally bad practice
+to commit files that do not pass tests but if it is necessary (e.g., for a local
+commit that will be rebased later), you can run ``git commit --no-verify`` to
+skip pre-commit checks.
```

Comment:
What happens if I run `git commit` outside the venv/conda environment that contains the `pre-commit` package?

Reply:
`pre-commit install` adds a git hook that uses the absolute path of whatever python interpreter was used to run the command. So if the install is run within an environment, the hooks run via `git commit` will just use that environment's python install, which works even if the environment isn't active.

If pre-commit is installed to the system (via `apt` or `pip` outside of an environment), it just uses's the system's python.

Reply:
> which works even if the environment isn't active

Nice. I didn't know that.